#! /bin/sh /usr/share/dpatch/dpatch-run
## 02_traversal_directory.dpatch by Jose Carlos Medeiros <debian@psabs.com.br>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: patch to solve problem with "directory traversal bug" CVE id CAN-2005-2349

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch ${2:+-d $2}}"

if [ $# -lt 1 ]; then
  echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
  exit 1
fi

case "$1" in
  -patch) patch $patch_opts -p1 < $0;;
  -unpatch) patch $patch_opts -p1 -R < $0;;
  *)
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1;;
esac

exit 0

@DPATCH@
diff -urNad zoo-2.10/portable.c /tmp/dpep.W7Aoaj/zoo-2.10/portable.c
--- zoo-2.10/portable.c	2005-07-22 15:34:22.000000000 -0300
+++ /tmp/dpep.W7Aoaj/zoo-2.10/portable.c	2005-07-25 14:43:57.757855384 -0300
@@ -364,6 +364,41 @@
       show_dir(direntry);
    }
 #endif
+   /* #########################################################################
+    * 
+    *  THIS CODE WAS WRITTEN TO SOLVE PROBLEM WITH DIRECTORY TRAVERSAL SECURITY
+    *  BUG (CVE id CAN-2005-2349).
+    *  
+    *  ########################################################################
+    */
+   char *p;
+   /* take off '../'   */
+   while ((p = strstr( direntry->dirname, "../" )) != NULL) {
+      while (*(p+3) != '\0') {
+        *p = *(p + 3);
+        p++;
+      }
+      *p = *(p+3); /* move last null */
+      //printf("zoo: skipped \"../\" path component in '%s'\n", direntry->dirname);
+   }
+   /* take off  '/'  */
+   if ( direntry->dirname[0] == '/' ) {
+      p = direntry->dirname;
+      while (*p != '\0') {
+        *p = *(p + 1);
+        p++;
+      }
+      *p = *(p+1); /* move last null */
+      //printf("zoo: skipped \"/\" path component in '%s'\n", direntry->dirname);
+   }
+   /* direntry->dirlen = strlen(direntry->dirname); */
+
+   /* ##################################################################
+    * 
+    * END
+    *
+    * ###################################################################
+   */
    return (0);
 }
 
